onnxruntime/contrib_ops/cpu/quantization/moe_quantization_cpu.cc

```diff
@@ -152,15 +152,31 @@ Status QMoE::QuantizedMoEImpl(OpKernelContext* context,
 
   auto thread_fc1_buffers = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(num_threads * moe_params.inter_size * (is_swiglu ? 2 : 1)));
   auto thread_fc2_buffers = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(num_threads * moe_params.hidden_size));
-  auto thread_results = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(num_threads * moe_params.num_rows * moe_params.hidden_size));
 
-  const int64_t max_bias_size = std::max(moe_params.inter_size * (is_swiglu ? 2 : 1), moe_params.hidden_size);
-  auto thread_bias_buffers = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(num_threads * max_bias_size));
+  // Allocate a single output buffer instead of per-thread buffers
+  auto output_float = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(total_output_size));
+  std::fill_n(output_float.get(), total_output_size, 0.0f);
 
-  // Prepare float buffers for input data
+  // Prepare float buffers for input data and biases
   auto input_float = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(moe_params.num_rows * moe_params.hidden_size));
   auto router_probs_float = IAllocator::MakeUniquePtr<float>(allocator, static_cast<size_t>(moe_params.num_rows * moe_params.num_experts));
 
+  // Pre-convert bias tensors to float (if they exist)
+  const int64_t fc1_bias_size = is_swiglu ? 2 * moe_params.inter_size : moe_params.inter_size;
+  const int64_t fc2_bias_size = moe_params.hidden_size;
+
+  // Allocate buffers for converted biases
+  std::unique_ptr<float[]> fc1_bias_float;
+  std::unique_ptr<float[]> fc2_bias_float;
+
+  if (fc1_bias_data) {
+    fc1_bias_float = std::make_unique<float[]>(static_cast<size_t>(moe_params.num_experts * fc1_bias_size));
```

Comment:
Please use ORT allocator. The Arena could help performance than allocating in heap.